### PR TITLE
Use queue.Queue, new /status,/volume/ & /speed/ endpoints, new full selection mode, call ffplay and piper internally and a bunch of cosmetic changes

### PR DESCRIPTION
**Pull request by kratuvid**
Bunch of opinionated improvements. At your discretion

Cosmetic changes:
1. Consistent variable and function names
2. Better English at some places
3. Structural organization

Others:
1. Removed argparse in requirements.txt. It's builtin Python
3. Removed the usage of shell scripts. The program now internally calls ffplay and piper
4. queue (pass_queue) is now queue.Queue instead of a list. Because it meant for the exact purpose. https://docs.python.org/3/library/queue.html
5. The program can now process text in additional full selection mode (--full_selection). Just pass the whole clipboard contents into piper, instead of chunking by sentences. Slightly better sentence flow while reading. Tell me if it's not
6. Option to adjust playback sample rate in case the user is using models that output at difference sample rates. Example, low quality models output at 16000 Hz instead of 22050 Hz. https://github.com/rhasspy/piper/blob/master/TRAINING.md
7. Program can now fatally exit when needed
8. New /status API endpoint for basic stats
9. Dynamically change volume and speed using new endpoints

### CODE REVIEW

**Comment by dipta10** [comment]
Hi @kratuvid , thank you for the PR.

I agree that reading the whole clipboard instead of chunking provides better sentence flow. However, it may be slower when a large amount of text is selected. My original intent was to make it faster and therefore I used chunking.
But we can make full clipboard selection the default, I think that makes more sense.

File: main.py [hunk 1]
```diff
@@ -1,173 +1,267 @@
...
-            queue.append(file_name)
+        if not parsed.full_selection:
+            tokens = text.split(". ")
+            while tokens:
+                text = tokens[0].strip() + "."
+                tokens = tokens[1:]
+                text = sanitize_text(text)
+
+                out = generate_audio(text)
+                if out != None:
```
Comment (dipta10) [inline main.py]:
```suggestion
                if out is not None:
```
This is more pythonic. If you have some time we can change it accordingly.

File: main.py [hunk 2]
```diff
@@ -1,173 +1,267 @@
...
-                    f'"{text}"',
-                    parser.parse_args().model,
-                    parser.parse_args().model_config,
-                ]
-            )
-            process.wait()
-            queue.append(file_name)
+        if not parsed.full_selection:
+            tokens = text.split(". ")
+            while tokens:
```
Comment (dipta10) [inline main.py]: There is an issue when the chunking method is applied to a large body of text separated by periods.

If a stop request is received during the audio generation process, the queue is cleared correctly. However, this function continues to add the remaining unprocessed chunks to the queue. These unprocessed chunks of sentences are then read on the next read request.

```suggestion
            while tokens and not stop_event.is_set():            
```

I thinks this should solve the issue.

**Review by dipta10 — CHANGES_REQUESTED** [review]

File: main.py (hunk 2 above)
Comment (kratuvid) [reply main.py]: Yeah it's an edge case. Commiting

File: main.py (hunk 1 above)
Comment (kratuvid) [reply main.py]: Did not know about that. Changing to is in all locations

**Comment by kratuvid** [comment]
>But we can make full clipboard selection the default, I think that makes more sense

I thought that too. Making it the default

>One comes to value this plus health, when he sees that all difficulties vanish before it. A timid man listening to the alarmists in Congress, and in the newspapers, and observing the profligacy of party, — sectional interests urged with a fury which shuts its eyes to consequences, with a mind made up to desperate extremities, ballot in one hand, and rifle in the other, — might easily believe that he and his country have seen their best days, and he hardens himself the best he can against the coming ruin. But, after this has been foretold with equal confidence fifty times, and government six per cents have not declined a quarter of a mill, he discovers that the enormous elements of strength which are here in play, make our politics unimportant. Personal power, freedom, and the resources of nature strain every faculty of every citizen. We prosper with such vigor, that, like thrifty trees, which grow in spite of ice, lice, mice, and borers, so we do not suffer from the profligate swarms that fatten on the national treasury. The huge animals nourish huge parasites, and the rancor of the disease attests the strength of the constitution. The same energy in the Greek Demos drew the remark, that the evils of popular government appear greater than they are; there is compensation for them in the spirit and energy it awakens. The rough and ready style which belongs to a people of sailors, foresters, farmers, and mechanics, has its advantages. Power educates the potentate. As long as our people quote English standards they dwarf their own proportions. A Western lawyer of eminence said to me he wished it were a penal offence to bring an English law-book into a court in this country, so pernicious had he found in his experience our deference to English precedent. The very word 'commerce' has only an English meaning, and is pinched to the cramp exigencies of English experience. The commerce of rivers, the commerce of railroads, and who knows but the commerce of air-balloons, must add an American extension to the pond-hole of admiralty. As long as our people quote English standards, they will miss the sovereignty of power; but let these rough riders, — legislators in shirt-sleeves, — Hoosier, Sucker, Wolverine, Badger, — or whatever hard head Arkansas, Oregon, or Utah sends, half orator, half assassin, to represent its wrath and cupidity at Washington, — let these drive as they may; and the disposition of territories and public lands, the necessity of balancing and keeping at bay the snarling majorities of German, Irish, and of native millions, will bestow promptness, address, and reason, at last, on our buffalo-hunter, and authority and majesty of [manners](https://emersoncentral.com/texts/essays-second-series/manners/). The instinct of the people is right. Men expect from good whigs, put into office by the respectability of the country, much less skill to deal with Mexico, Spain, Britain, or with our own malcontent members, than from some strong transgressor, like Jefferson, or Jackson, who first conquers his own government, and then uses the same genius to conquer the foreigner. The senators who dissented from Mr. Polk's Mexican war, were not those who knew better, but those who, from political position, could afford it; not Webster, but Benton and Calhoun.

It takes 3-5 seconds on my Ryzen 5 5600H to play a fairly big paragraph like this. I don't think it is much. Can't comment on slower hardware

**Comment by kratuvid** [comment]
i think my commits are too intrusive. it's better to stay on my fork. closing

**Comment by kratuvid** [comment]
i think my commits are too intrusive. it's better to stay on my fork. closing

**Comment by dipta10** [comment]
Hey @kratuvid , your commits do not seem intrusive to me. If anything it made the shitty code way better with lots of options. If you have some time to create a PR again, I would be delighted to test it in i3 and merge it.

**Pull request closed without merging**